com.unity.netcode.gameobjects/Runtime/Core/SnapshotSystem.cs

```diff
@@ -59,141 +65,102 @@ internal struct SnapshotSpawnCommand
-        private const int k_MaxVariables = 2000;
-        internal int SpawnsBufferCount { get; private set; } = 100;
-        internal int DespawnsBufferCount { get; private set; } = 100;
-
-        private const int k_BufferSize = 30000;
-
-        private NetworkManager m_NetworkManager = default;
+        private NetworkManager m_NetworkManager;
+        private NetworkTickSystem m_NetworkTickSystem;
 
-        // by clientId
         private Dictionary<ulong, ClientData> m_ClientData = new Dictionary<ulong, ClientData>();
         private Dictionary<ulong, ConnectionRtt> m_ConnectionRtts = new Dictionary<ulong, ConnectionRtt>();
 
-        private bool m_UseSnapshotDelta;
-        private bool m_UseSnapshotSpawn;
-        private int m_SnapshotMaxSpawnUsage;
-        private NetworkTickSystem m_NetworkTickSystem;
-
+        // The tick we're currently processing (or last we processed, outside NetworkUpdate())
         private int m_CurrentTick = NetworkTickSystem.NoTick;
 
-        internal byte[] MainBuffer = new byte[k_BufferSize]; // buffer holding a snapshot in memory
-        internal byte[] RecvBuffer = new byte[k_BufferSize]; // buffer holding the received snapshot message
-
-        internal IndexAllocator Allocator;
-
-        internal Entry[] Entries = new Entry[k_MaxVariables];
-        internal int LastEntry = 0;
-
+        // The container for the spawn commands received by the user. This part can be written as-is to the message
         internal SnapshotSpawnCommand[] Spawns;
+        // Information about Spawns. Entries are matched by index
+        internal SnapshotSpawnCommandMeta[] SpawnsMeta;
         internal int NumSpawns = 0;
 
+        // The container for the despawn commands received by the user. This part can be written as-is to the message
         internal SnapshotDespawnCommand[] Despawns;
+        // Information about Despawns. Entries are matched by index
+        internal SnapshotDespawnCommandMeta[] DespawnsMeta;
         internal int NumDespawns = 0;
 
+        private static int s_DebugNextId = 0;
+        private int m_DebugMyId = 0;
+
+        // Local state. Stores which spawns and despawns were applied locally
         // indexed by ObjectId
         internal Dictionary<ulong, int> TickAppliedSpawn = new Dictionary<ulong, int>();
         internal Dictionary<ulong, int> TickAppliedDespawn = new Dictionary<ulong, int>();
 
+        // Settings
         internal bool IsServer { get; set; }
         internal bool IsConnectedClient { get; set; }
         internal ulong ServerClientId { get; set; }
         internal List<ulong> ConnectedClientsId { get; } = new List<ulong>();
-        internal MockSendMessage MockSendMessage { get; set; }
-        internal MockSpawnObject MockSpawnObject { get; set; }
-        internal MockDespawnObject MockDespawnObject { get; set; }
+        internal SendMessageDelegate SendMessage { get; set; }
```

Comment:
consider Events.

Reply:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/event

Reply:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#names-of-common-types

Reply:
doc> DO NOT add the suffix "Delegate" to a delegate.

LOL. I like your style. Thanks for pointing me to the documentation. Not sure what to call it though.

Reply:
for delegate definition:
```cs
internal delegate int SendMessageEventHandler(SnapshotDataMessage message, ulong clientId);
```

for event field declaration:
```cs
internal event SendMessageEventHandler OnSendMessage { get; set; }
```